File: ruby/check.rb
# Driver interoperability check for ExecDB's pgwire (spec §8, phase 4
# follow-up): Ruby's pg gem, against a running ExecDB instance already
# seeded with table t(a INTEGER) (seeded by run-all.sh below).
#
# pg's exec_params sends real Extended Query parameter binding (Parse/
# Bind), unlike PHP's PDO_PGSQL default (php/check.php) --
# even though both are, like psycopg2, libpq wrappers under the hood, this
# default usage pattern differs and is worth its own check.
require 'pg'

if ARGV.length != 1
  warn 'usage: check.rb <conninfo>'
  exit 2
end

conn = PG.connect(ARGV[0])

res = conn.exec_params('SELECT $1', [1])
one = res.values[0][0]
raise "SELECT $1 (1) returned #{one.inspect}" unless one.to_s == '1'

res = conn.exec_params('SELECT $1', [3.5])
f = res.values[0][0]
raise "SELECT $1 (3.5) returned #{f.inspect}" unless f.to_f == 3.5

res = conn.exec_params('SELECT $1', ['hello'])
s = res.values[0][0]
raise "SELECT $1 ('hello') returned #{s.inspect}" unless s == 'hello'

# result_format 1 (binary): pg's default text format returns bytea as
# the literal "\x00ff"-style hex-escape string, not decoded into raw
# bytes -- unlike psycopg2, which decodes it automatically even in text
# mode (python/check.py). Requesting binary here checks
# ExecDB's binary bytea encoding instead of re-deriving pg's own text
# escaping.
res = conn.exec_params("SELECT x'00ff'", [], 1)
blob = res.values[0][0]
raise "SELECT x'00ff' returned #{blob.inspect}" unless blob == "\x00\xff".b

res = conn.exec('SELECT NULL')
null = res.values[0][0]
raise "SELECT NULL returned #{null.inspect}" unless null.nil?

# spec §2: DDL must be rejected via the external I/F, and pg must surface
# it as a PG::Error carrying SQLSTATE 42501 (spec §8).
begin
  conn.exec('CREATE TABLE ruby_should_not_exist(a INTEGER)')
  raise 'expected CREATE TABLE to be rejected via the external I/F'
rescue PG::Error => e
  code = e.result.error_field(PG::PG_DIAG_SQLSTATE)
  raise "expected SQLSTATE 42501, got #{code.inspect} (#{e.message})" unless code == '42501'
end

# A basic write/read round trip against the table run-all.sh seeded.
conn.exec_params('INSERT INTO t VALUES ($1)', [777010])
res = conn.exec_params('SELECT count(*) FROM t WHERE a = $1', [777010])
n = res.values[0][0].to_i
raise "expected count=1 after INSERT, got #{n}" unless n == 1

puts 'OK'

File: php/check.php
<?php
// Driver interoperability check for ExecDB's pgwire (spec §8, phase 4
// follow-up): PHP's PDO_PGSQL (via PDO), against a running ExecDB
// instance already seeded with table t(a INTEGER) (seeded by run-all.sh below).
//
// PDO_PGSQL defaults to "emulated prepares" (PDO::ATTR_EMULATE_PREPARES
// is true by default for this driver): parameter values are substituted
// into the SQL text on the client side and the whole statement is sent
// as plain text via Simple Query, not Extended Query's Parse/Bind. This
// is a genuinely different code path from every other verified driver
// (all of which use real Extended Query parameter binding by default),
// even though PDO_PGSQL, like psycopg2, is itself a libpq wrapper --
// worth keeping even though libpq's own wire handling was already
// exercised by psycopg2, precisely because this default differs.

if ($argc !== 2) {
    fwrite(STDERR, "usage: check.php <dsn>\n");
    exit(2);
}

$dsn = $argv[1];
$pdo = new PDO($dsn);
$pdo->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);

$stmt = $pdo->prepare("SELECT ?");
$stmt->execute([1]);
$one = $stmt->fetchColumn();
if ((string)$one !== "1") {
    fwrite(STDERR, "SELECT ? (1) returned " . var_export($one, true) . "\n");
    exit(1);
}

$stmt = $pdo->prepare("SELECT ?");
$stmt->execute([3.5]);
$f = $stmt->fetchColumn();
if ((float)$f !== 3.5) {
    fwrite(STDERR, "SELECT ? (3.5) returned " . var_export($f, true) . "\n");
    exit(1);
}

$stmt = $pdo->prepare("SELECT ?");
$stmt->execute(["hello"]);
$s = $stmt->fetchColumn();
if ($s !== "hello") {
    fwrite(STDERR, "SELECT ? ('hello') returned " . var_export($s, true) . "\n");
    exit(1);
}

$stmt = $pdo->query("SELECT NULL");
$null = $stmt->fetchColumn();
if ($null !== null) {
    fwrite(STDERR, "SELECT NULL returned " . var_export($null, true) . "\n");
    exit(1);
}

// spec §2: DDL must be rejected via the external I/F, and PDO must
// surface it as a PDOException carrying SQLSTATE 42501 (spec §8).
try {
    $pdo->exec("CREATE TABLE php_should_not_exist(a INTEGER)");
    fwrite(STDERR, "expected CREATE TABLE to be rejected via the external I/F\n");
    exit(1);
} catch (PDOException $e) {
    if ($e->getCode() !== "42501") {
        fwrite(STDERR, "expected SQLSTATE 42501, got " . var_export($e->getCode(), true) . " (" . $e->getMessage() . ")\n");
        exit(1);
    }
}

// A basic write/read round trip against the table run-all.sh seeded.
$stmt = $pdo->prepare("INSERT INTO t VALUES (?)");
$stmt->execute([777009]);
$stmt = $pdo->prepare("SELECT count(*) FROM t WHERE a = ?");
$stmt->execute([777009]);
$n = $stmt->fetchColumn();
if ((int)$n !== 1) {
    fwrite(STDERR, "expected count=1 after INSERT, got " . var_export($n, true) . "\n");
    exit(1);
}

echo "OK\n";
